In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import ipywidgets as widgets
from IPython.display import display

In [2]:
heart_data = pd.read_csv('HDP.csv')

In [3]:
heart_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [4]:
heart_data.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [5]:
heart_data.shape

(303, 14)

In [6]:
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [7]:
heart_data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [8]:

\
heart_data.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [9]:
heart_data['target'].value_counts()

target
1    165
0    138
Name: count, dtype: int64

In [10]:
X = heart_data.drop(columns='target', axis=1)
Y = heart_data['target']

In [11]:
print(X)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1        0   0     2  
2        2   0    

In [12]:
print(Y)

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64


In [13]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [14]:
print(X.shape, X_train.shape, X_test.shape)

(303, 13) (242, 13) (61, 13)


In [15]:
model = LogisticRegression()

In [16]:
model.fit(X_train, Y_train)

C:\Users\Likith G S\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [17]:
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [18]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.8512396694214877


In [19]:
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [20]:
print('Accuracy on Test data : ', test_data_accuracy)

Accuracy on Test data :  0.819672131147541


In [21]:
age = widgets.FloatText(description="Age:")
sex = widgets.Dropdown(
    options=[('Male', 0), ('Female', 1)],
    description='Sex:'
)
cp = widgets.FloatText(description="CP:")
trestbps = widgets.FloatText(description="Trestbps:")
chol = widgets.FloatText(description="Chol:")
fbs = widgets.FloatText(description="FBS:")
restecg = widgets.FloatText(description="Restecg:")
thalach = widgets.FloatText(description="Thalach:")
exang = widgets.Dropdown(
    options=[('No', 0), ('Yes', 1)],
    description='Exang:'
)
oldpeak = widgets.FloatText(description="Oldpeak:")
slope = widgets.FloatText(description="Slope:")
ca = widgets.FloatText(description="CA:")
thal = widgets.FloatText(description="Thal:")

button = widgets.Button(description="Predict")
output = widgets.Output()

In [22]:
def on_button_click(b):
    with output:
        input_data = [
            age.value,
            sex.value,
            cp.value,
            trestbps.value,
            chol.value,
            fbs.value,
            restecg.value,
            thalach.value,
            exang.value,
            oldpeak.value,
            slope.value,
            ca.value,
            thal.value  # Add this line for thal
        ]

        # Convert input data to a numpy array and reshape it
        input_data_as_numpy_array = np.asarray(input_data)
        input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)

        # Make a prediction
        prediction = model.predict(input_data_reshaped)

        if prediction[0] == 0:
            result = 'The Person does not have a Heart Disease'
        else:
            result = 'The Person has Heart Disease'

        # Clear previous output and display the result
        output.clear_output()
        print(result)

button.on_click(on_button_click)

**Heart Disease Prediction Form: Field Descriptions**

1. **age**: Type in the patient's age.

2. **sex**: Choose the patient's sex (Male or Female) from a dropdown list.

3. **cp**: Enter the type of chest pain the patient experiences.

4. **trestbps**: Type in the patient's resting blood pressure (measured in mm Hg).

5. **chol**: Enter the patient's cholesterol level (measured in mg/dl).

6. **fbs**: Indicate whether the patient’s fasting blood sugar level is above 120 mg/dl (No or Yes).

7. **restecg**: Enter the result of the patient's resting electrocardiogram (Normal or Abnormal).

8. **thalach**: Type in the maximum heart rate the patient achieved during exercise.

9. **exang**: Choose whether the patient had exercise-induced angina (No or Yes).

10. **oldpeak**: Enter the level of depression experienced during exercise compared to rest.

11. **slope**: Enter the slope of the ST segment on the heart's exercise test (Upsloping, Flat, or Downsloping).

12. **ca**: Type in the number of major blood vessels that were colored during a fluoroscopy test.


In [23]:
display(age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, button, output)

FloatText(value=0.0, description='Age:')

Dropdown(description='Sex:', options=(('Male', 0), ('Female', 1)), value=0)

FloatText(value=0.0, description='CP:')

FloatText(value=0.0, description='Trestbps:')

FloatText(value=0.0, description='Chol:')

FloatText(value=0.0, description='FBS:')

FloatText(value=0.0, description='Restecg:')

FloatText(value=0.0, description='Thalach:')

Dropdown(description='Exang:', options=(('No', 0), ('Yes', 1)), value=0)

FloatText(value=0.0, description='Oldpeak:')

FloatText(value=0.0, description='Slope:')

FloatText(value=0.0, description='CA:')

FloatText(value=0.0, description='Thal:')

Button(description='Predict', style=ButtonStyle())

Output()

In [25]:
heart_data.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [28]:
import pickle
# Save the model using pickle at the end of the script
with open('heart_disease_model.pkl', 'wb') as file:
    pickle.dump(model, file)
    print("Model saved as 'heart_disease_model.pkl'")

Model saved as 'heart_disease_model.pkl'
